### PR TITLE
[crypto] Consolidate the OTBN driver and make it stateless.

### PR DESCRIPTION
**Pull request by jadephilipoom**
Previously, the functional OTBN driver for cryptolib was split between drivers/otbn.c and impl/otbn_util.c, with the latter providing a few basic utility functions in terms of the former. The otbn_util.c version also included a state struct which simply stated which app was loaded into OTBN (if any).

This commit completely removes impl/otbn_util.c. It removes the state struct and moves some functions to drivers/otbn.c while eliminating others (for instance, `otbn_copy_data_to_otbn()` is replaced with `otbn_dmem_write()`, since they have a very similar interface). Accordingly, some of existing functions (including `otbn_dmem_write`) are adjusted to be more intuitive to use (e.g. by using an `otbn_addr_t` instead of a raw byte offset).

The motivations for this change are:
- putting all the source code in one place so the driver easier to understand, use, and test
- removing the state struct, which doesn't quite fit with the (stateless) cryptolib
- un-expose several functions/constants from the driver that were previously needed by `otbn_util.c` but shouldn't be needed by non-driver code

@alphan it's possible to do a similar refactor for the `silicon_creator` driver; WDYT?

### CODE REVIEW

File: sw/device/lib/crypto/drivers/otbn.c [hunk 1]
```diff
@@ -194,3 +229,56 @@ otbn_error_t otbn_set_ctrl_software_errs_fatal(bool enable) {
...
+  const size_t imem_num_words = app.imem_end - app.imem_start;
+  const size_t data_num_words = app.dmem_data_end - app.dmem_data_start;
+
+  OTBN_RETURN_IF_ERROR(otbn_imem_sec_wipe());
+  // IMEM always starts at zero.
+  otbn_addr_t imem_start_addr = 0;
+  OTBN_RETURN_IF_ERROR(
+      otbn_imem_write(imem_num_words, app.imem_start, imem_start_addr));
+
+  OTBN_RETURN_IF_ERROR(otbn_dmem_sec_wipe());
```
Comment (jrvanwhy) [inline sw/device/lib/crypto/drivers/otbn.c:277]: I realize this is a refactoring PR and therefore this question is out of scope, but is there a reason the IMEM write was performed before the DMEM wipe? I'm assuming that IMEM is instruction memory and DMEM is data memory. It seems like a successful instruction write followed by a failed data wipe would leave OTBN in kindof a dangerous state.

**Review by jrvanwhy — APPROVED** [review]
The refactoring seems good to my new-to-this-codebase eye.

File: sw/device/lib/crypto/drivers/otbn.c [hunk 2]
```diff
@@ -35,10 +35,41 @@ ASSERT_ERR_BIT_MATCH(kOtbnErrBitsLifecycleEscalation,
...
+ * OTBN status
+ */
+typedef enum otbn_status {
+  kOtbnStatusIdle = 0x00,
+  kOtbnStatusBusyExecute = 0x01,
+  kOtbnStatusBusySecWipeDmem = 0x02,
+  kOtbnStatusBusySecWipeImem = 0x03,
+  kOtbnStatusBusySecWipeInt = 0x04,
+  kOtbnStatusLocked = 0xFF,
+} otbn_status_t;
```
Comment (alphan) [inline sw/device/lib/crypto/drivers/otbn.c:72]: Are there auto-generated macros for these values?

File: sw/device/lib/crypto/drivers/otbn.c (hunk 1 above)
Comment (alphan) [reply sw/device/lib/crypto/drivers/otbn.c:277]: Good point, thanks for the detailed review @jrvanwhy. Are you concerned about potentially exposing stale data in DMEM? That clearing should be done by the previous client of OTBN. The goal of the data wipe here is to make sure that the initial values of the DMEM are not easily predictable, making FI attacks on the dmem_write operation below harder to control and reproduce. PLMK if you had something else in mind.

**Review by alphan — APPROVED** [review]
Thanks @jadephilipoom !

> @alphan it's possible to do a similar refactor for the silicon_creator driver; WDYT?

I think we should do the same cleanup in silicon_creator code as well. Can you also take a look at `sw/device/lib/runtime/otbn.c` and see if that can be removed as well?

File: sw/device/lib/crypto/drivers/otbn.c (hunk 1 above)
Comment (alphan) [reply sw/device/lib/crypto/drivers/otbn.c:277]: Discussed this with @jadephilipoom offline and we decided to move the dmem_sec_wipe after imem_sec_wipe.

Comment (jrvanwhy) [reply sw/device/lib/crypto/drivers/otbn.c:277]: > Are you concerned about potentially exposing stale data in DMEM?

Yes, that was my concern.

Obviously you two have much better context on this code than I do so I have no further input.

File: sw/device/lib/crypto/drivers/otbn.c (hunk 2 above)
Comment (jadephilipoom) [reply sw/device/lib/crypto/drivers/otbn.c:72]: Apparently not. Created https://github.com/lowRISC/opentitan/issues/16754 to track this.

**Comment by jadephilipoom** [comment]
> I think we should do the same cleanup in silicon_creator code as well. Can you also take a look at `sw/device/lib/runtime/otbn.c` and see if that can be removed as well?

Created https://github.com/lowRISC/opentitan/issues/16789 to track these follow-ups.

**Pull request merged**